### PR TITLE
Add exploit-shape regression vectors for issue #39 findings 1-5

### PR DESCRIPTION
**Pull request by maximmasiutin**
## Summary

Extends `SoftInvalidFreeMemTest.dpr` with controlled exploit-shape regression
vectors that verify the foreign-pointer guards added by PRs #60 through #64
stay in place across 32-bit ASM, 64-bit ASM, and Pure Pascal builds.

## Vectors

Each vector constructs a guarded region (writable page + PAGE_NOACCESS /
PROT_NONE guard page), writes a caller-controlled fake header at P-8, then
invokes FreeMem or ReallocMem. A regression that removes a guard manifests
as EAccessViolation, tripwire corruption inside the writable page, or a
non-nil / non-zero return.

| Vector | Finding | PR  | Target guard |
|--------|---------|-----|--------------|
| Finding1_ReallocMemSmallForeign          | 1 | #63 | ASM FastReallocMem small-block pool-pointer |
| Finding1b_ReallocMemSmallMappedGarbageBlockType | 1 | #63 | ASM FastReallocMem SmallBlockType bounds |
| Finding2_ReallocMemMediumForeign         | 2 | #64 | ASM FastReallocMem medium-block size |
| Finding3_FreeMemLargeForeign             | 3 | #62 | ASM FreeMem large-block size/granularity |
| Finding3b_FreeMemLargeNonGranularity     | 3 | #62 | ASM FreeMem large-block granularity |
| Finding4_FreeMemSmallBlockTypeClearFillChar | 4 | #61 | ASM FreeMem SmallBlockType guard order |
| Finding5_FreeMemMediumSizeClearFillChar  | 5 | #60 | FreeMem medium size guard order (Pascal + ASM) |
| Finding5b_FreeMemMediumUndersized        | 5 | #60 | FreeMem undersized medium header |

Also adds a ReallocMem cycle regression test.

## CI matrix additions

Adds `-dAlwaysClearFreedMemory` variants of the existing SoftInvalidFreeMemTest
invocations to exercise the ClearSmall / ClearMedium FillChar guard ordering,
which is only compiled when `ClearSmallAndMediumBlocksInFreeMem` is active:

- Windows 32-bit ASM + AlwaysClearFreedMemory
- Windows 64-bit ASM + AlwaysClearFreedMemory
- Windows 32-bit Pure Pascal + AlwaysClearFreedMemory
- Windows 64-bit Pure Pascal + AlwaysClearFreedMemory
- Linux 64-bit ASM + AlwaysClearFreedMemory
- Linux 64-bit Pure Pascal + AlwaysClearFreedMemory

## Bonus

Silences three FPC `read but nowhere assigned` warnings for
`FastMMCpuFeaturesB`, `FastMMCpuSmallestMonitorLineSize`, and
`FastMMCpuLargestMonitorLineSize`, which are declared under `USE_CPUID`
but assigned only under `EnableWaitPKG`.

## Test plan

- [ ] CI SoftInvalidFreeMemTest passes under all 12 matrix configs (6 previous + 6 new AlwaysClearFreedMemory).
- [ ] Locally verified on Windows x86_64 FPC: ASM-SC-64 and PAS-SC-64 both report 14 passed, 0 failed.
- [ ] No new warnings introduced.

## Summary by CodeRabbit

* **Tests**
  * Extended CI/CD pipeline to run additional test variants across all platform and architecture combinations with enhanced memory initialization coverage.

* **Bug Fixes**
  * Fixed uninitialized variables in memory manager initialization that could cause compiler warnings.

* **New Features**
  * Added new test procedures for validating memory allocation edge cases and security scenarios.

### CODE REVIEW

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 1]
```diff
@@ -357,8 +357,11 @@ function fpmmap(addr: Pointer; len: PtrUInt; prot: Integer;
...
 function fpmunmap(addr: Pointer; len: PtrUInt): Integer; cdecl;
   external 'c' name 'munmap';
+function fpmprotect(addr: Pointer; len: PtrUInt; prot: Integer): Integer; cdecl;
+  external 'c' name 'mprotect';
 
 const
+  PROT_NONE  = 0;
   PROT_READ  = 1;
   PROT_WRITE = 2;
   MAP_PRIVATE   = $02;
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr:367]: MAP_ANONYMOUS is already defined at line 368 of this file. False positive.

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 2]
```diff
@@ -401,6 +404,571 @@ begin
...
+  if Region = nil then Exit;
+  { Protect the second page as no-access. }
+  if not VirtualFree(Pointer(PByte(Region) + (GUARD_REGION_SIZE shr 1)),
+                     GUARD_REGION_SIZE shr 1, $4000 { MEM_DECOMMIT }) then
+  begin
+    VirtualFree(Region, 0, MEM_RELEASE);
+    Exit;
+  end;
+  Base := Region;
+  P := Pointer(PByte(Region) + 8);
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr:465]: The guard fires at P-SizeOf(Pointer) before any P-16 dereference. FastMM reads the first header word at P-SizeOf(Pointer) for all block types; the IsLargeBlockFlag check rejects the foreign pointer before the large-block header at P-16 is accessed. P at Base+SizeOf(Pointer) is correct.

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 3]
```diff
@@ -401,6 +404,571 @@ begin
...
+    MAP_PRIVATE or MAP_ANONYMOUS, -1, 0);
+  if (Region = Pointer(-1)) or (Region = nil) then Exit;
+  if fpmprotect(Pointer(PByte(Region) + (GUARD_REGION_SIZE shr 1)),
+                GUARD_REGION_SIZE shr 1, PROT_NONE) <> 0 then
+  begin
+    fpmunmap(Region, GUARD_REGION_SIZE);
+    Exit;
+  end;
+  Base := Region;
+  P := Pointer(PByte(Region) + 8);
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr:480]: Same rationale as above -- guard fires at P-SizeOf(Pointer) before P-16 is accessed.

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 4]
```diff
@@ -401,6 +404,571 @@ begin
...
+    begin
+      Result := False;
+      Exit;
+    end;
+end;
+
+procedure WriteFakeHeader(P: Pointer; Value: NativeUInt);
+begin
+  PNativeUInt(PByte(P) - SizeOf(Pointer))^ := Value;
+end;
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr:515]: Not needed. FastMM reads the header at P-SizeOf(Pointer) for all block types. The guard rejects foreign pointers at that offset before any large-block P-16 access.

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 5]
```diff
@@ -401,6 +404,571 @@ begin
...
+  ExcClass: string;
+begin
+  if not AllocGuardedRegion(Base, P) then
+  begin
+    TestFail(TestName, 'AllocGuardedRegion failed');
+    Exit;
+  end;
+  try
+    { Header $04: IsLargeBlockFlag set, masked size = 0. }
+    WriteFakeHeader(P, $04);
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr:682]: Same rationale -- guard fires at P-SizeOf(Pointer) header check.

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 6]
```diff
@@ -401,6 +404,571 @@ begin
...
+  Outcome: TFreeOutcome;
+  ExcClass: string;
+begin
+  if not AllocGuardedRegion(Base, P) then
+  begin
+    TestFail(TestName, 'AllocGuardedRegion failed');
+    Exit;
+  end;
+  try
+    WriteFakeHeader(P, $00008004);
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr:855]: Same rationale as finding 3 reply.

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 7]
```diff
@@ -401,6 +404,571 @@ begin
...
+  Result := False;
+  Base := nil;
+  P := nil;
+{$IFDEF MSWINDOWS}
+  { Allocate 2 pages: first PAGE_READWRITE (usable), second PAGE_NOACCESS (guard). }
+  Region := VirtualAlloc(nil, GUARD_REGION_SIZE, MEM_COMMIT, PAGE_READWRITE);
+  if Region = nil then Exit;
+  { Protect the second page as no-access. }
+  if not VirtualFree(Pointer(PByte(Region) + (GUARD_REGION_SIZE shr 1)),
+                     GUARD_REGION_SIZE shr 1, $4000 { MEM_DECOMMIT }) then
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr]: Fixed. Comment now says decommitted instead of PAGE_NOACCESS.

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 8]
```diff
@@ -401,6 +404,571 @@ begin
...
+// Controlled exploit-shape vectors for issue #39 findings 1-5 (PRs #60-#64).
+//
+// Each vector constructs a guarded region (writable page + PAGE_NOACCESS /
+// PROT_NONE guard page), places a user pointer P at offset 8 so the fake
+// header at P-8 lives in the writable page, writes a specific header value
+// that triggers one finding's pre-patch failure mode, then calls FreeMem or
+// ReallocMem. On patched code the guard fires and returns 0/nil without
+// touching the guard page. On unpatched code FreeMem/ReallocMem would read
+// or write beyond the writable page (AV) or dereference an invalid pool /
+// BlockType pointer (AV) or execute an unbounded FillChar.
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr]: Fixed. Comment now says P-SizeOf(Pointer).

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 9]
```diff
@@ -401,6 +404,571 @@ begin
...
+end;
+{$ENDIF}
+{$ENDIF}
+
+// -----------------------------------------------------------------------------
+// Finding 5 (PR #60, PR-SEC-02): FreeMem ClearSmallAndMedium FillChar runs
+// before the medium-block size check. Header $00100002: IsMediumBlockFlag set,
+// masked size = $00100000 (1MB). Unpatched: FillChar(P, 1MB - 8, 0) writes
+// past the 4KB writable region into the guard page. Patched: size check
+// rejects before FillChar ($100000 > MediumBlockPoolSize - header size).
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr]: Fixed. Comment updated to match $10000002 / 256MB.

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 10]
```diff
@@ -401,6 +404,571 @@ begin
...
+  begin
+    FreeGuardedRegion(Base);
+    TestFail(TestName, 'HeapAlloc returned nil');
+    Exit;
+  end;
+  try
+    { Header = address of Q (>= $10000, low 3 bits clear because HeapAlloc
+      returns 16-byte aligned addresses). Bypasses the pool-pointer guard
+      but has an invalid BlockType when the ASM reads [Q + BlockType_offset]. }
+    WriteFakeHeader(P, NativeUInt(Q));
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr:738]: Fixed. Added FillChar(Q^, 64, $A5).

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 11]
```diff
@@ -401,6 +404,571 @@ begin
...
+  Q2 := HeapAlloc(Heap, 0, 64);
+  if Q2 = nil then
+  begin
+    FreeGuardedRegion(Base);
+    TestFail(TestName, 'HeapAlloc returned nil');
+    Exit;
+  end;
+  try
+    { Header = Q2 address (>= $10000, low bits clear). Bypasses pool-pointer
+      guard but Q2's first 8 bytes are garbage, read as SmallBlockType pointer. }
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr]: Fixed. Added FillChar(Q2^, 64, $A5).

File: .github/workflows/main.yaml [hunk 12]
```diff
@@ -290,6 +290,23 @@ jobs:
           fpc -B -Mdelphi -Tlinux -Px86_64 -Co -Cr -dSoftInvalidFreeMem -dDontUseASMVersion -dIgnoreMemoryAllocatedBefore -O4 SoftInvalidFreeMemTest.dpr
           ./SoftInvalidFreeMemTest
 
+      # SoftInvalidFreeMem + AlwaysClearFreedMemory. Exercises the ClearSmall
+      # and ClearMedium FillChar guard ordering (findings 4 and 5, PRs #60/#61)
+      # which are only compiled when ClearSmallAndMediumBlocksInFreeMem is on.
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply .github/workflows/main.yaml]: Fixed. Comment now says only Finding 5 runs on Linux; Finding 4 is MSWINDOWS-only.

File: .github/workflows/main.yaml [hunk 13]
```diff
@@ -774,6 +791,39 @@ jobs:
           & '${{ env.fpc }}' @('-B', '-Mdelphi', '-Twin32', '-Pi386', '-Co', '-Cr', '-dSoftInvalidFreeMem', '-dDontUseASMVersion', '-dIgnoreMemoryAllocatedBefore', '-dDisableMemoryLeakReporting', '-O4', 'SoftInvalidFreeMemTest.dpr')
           & '.\SoftInvalidFreeMemTest.exe'
 
+      # SoftInvalidFreeMem + AlwaysClearFreedMemory variants. Exercise the
+      # ClearSmall / ClearMedium FillChar guard ordering (issue #39 findings
+      # 4 and 5, PRs #60 / #61). These code paths only compile when
+      # ClearSmallAndMediumBlocksInFreeMem is active. Covers 32-bit ASM,
+      # 64-bit ASM, and Pure Pascal paths.
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply .github/workflows/main.yaml:798]: Windows runs both findings 4 and 5 so the current wording is accurate. No change needed.

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 14]
```diff
@@ -401,6 +404,571 @@ begin
...
+// (that would crash the CI runner). It observes two things on patched code:
+// (a) FreeMem/ReallocMem returns 0 / nil (SoftInvalidFreeMem path).
+// (b) The tripwire bytes inside the writable page at offset >= 16 stay at
+//     their initial $5A pattern, proving FillChar did not run.
+// A regression that removes a guard would manifest as AV (process crash,
+// recorded by CI as a test failure) or tripwire corruption.
+// =============================================================================
+
+const
+  GUARD_REGION_SIZE    = $2000; { 8KB writable + 4KB guard on Windows / 4KB + 4KB on Linux }
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr]: Fixed. Comment corrected to 4KB+4KB.

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 15]
```diff
@@ -401,6 +404,571 @@ begin
...
+  if not AllocGuardedRegion(Base, P) then
+  begin
+    TestFail(TestName, 'AllocGuardedRegion failed');
+    Exit;
+  end;
+  try
+    { Header $10000002: IsMediumBlockFlag set, size after mask = $10000000
+      (256MB) which exceeds MediumBlockPoolSize ($13FFF0). The pre-FillChar
+      size guard must reject this before FillChar runs. }
+    WriteFakeHeader(P, $10000002);
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr:788]: Fixed in same commit. Comment updated to $10000002 / 256MB.

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 16]
```diff
@@ -401,6 +404,571 @@ begin
...
+//     their initial $5A pattern, proving FillChar did not run.
+// A regression that removes a guard would manifest as AV (process crash,
+// recorded by CI as a test failure) or tripwire corruption.
+// =============================================================================
+
+const
+  GUARD_REGION_SIZE    = $2000; { 8KB writable + 4KB guard on Windows / 4KB + 4KB on Linux }
+  TRIPWIRE_PATTERN     = $5A;
+  TRIPWIRE_CHECK_START = 16;    { skip header+pointer bytes }
+  TRIPWIRE_CHECK_LEN   = 256;
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr]: Fixed in same commit.

File: Tests/Simple/SoftInvalidFreeMemTest.dpr (hunk 10 above)
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr:738]: Fixed in same commit. Added FillChar(Q^, 64, $A5).

**Comment by maximmasiutin** [comment]
@coderabbit full review

File: Tests/Simple/SoftInvalidFreeMemTest.dpr [hunk 17]
```diff
@@ -401,6 +404,579 @@ begin
...
+    TestFail(TestName, 'HeapAlloc returned nil');
+    Exit;
+  end;
+  { Deterministic fill: ensures [Q + BlockType_offset] is reliably outside
+    SmallBlockTypes regardless of heap state. }
+  FillChar(Q^, 64, $A5);
+  try
+    { Header = address of Q (>= $10000, low 3 bits clear because HeapAlloc
+      returns 16-byte aligned addresses). Bypasses the pool-pointer guard
+      but has an invalid BlockType when the ASM reads [Q + BlockType_offset]. }
```
[the remark this answers is not in the document]
Comment (maximmasiutin) [reply Tests/Simple/SoftInvalidFreeMemTest.dpr]: Fixed. Changed to 'at least 8-byte aligned'.
